### DISCUSSION TITLE
Building on MacOS 14.1

### DISCUSSION THREAD

**Discussion by DavidACosgrove** [rdkit/rdkit#6859] [Development]
In case anyone's interested, I've just updated my MacBookPro (Apple Silicon) to Sonoma, MacOS 14.1.  Previously it was on Ventura.  RDKit no longer builds using my standard CMake spell, with an error very early on compiling a boost header:
```
In file included from /Users/david/Projects/RDKit/WorktreesMyFork/BetterLasso/External/CoordGen/maeparser/MaeBlock.cpp:1:
In file included from /Users/david/Projects/RDKit/WorktreesMyFork/BetterLasso/External/CoordGen/maeparser/MaeBlock.hpp:3:
In file included from /Users/david/mambaforge/envs/RDKit.Build/include/boost/dynamic_bitset.hpp:15:
In file included from /Users/david/mambaforge/envs/RDKit.Build/include/boost/dynamic_bitset/dynamic_bitset.hpp:53:
In file included from /Users/david/mambaforge/envs/RDKit.Build/include/boost/functional/hash/hash.hpp:6:
/Users/david/mambaforge/envs/RDKit.Build/include/boost/container_hash/hash.hpp:131:33: error: no template named 'unary_function' in namespace 'std'; did you mean '__unary_function'?
        struct hash_base : std::unary_function<T, std::size_t> {};
                           ~~~~~^~~~~~~~~~~~~~
                                __unary_function
/Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__functional/unary_function.h:46:1: note: '__unary_function' declared here
using __unary_function = __unary_function_keep_layout_base<_Arg, _Result>;
^
```
If I edit the CMakeCache.txt to use the previous compiler (look for 14.0 and change it to 13.3), it goes fine.  This is using boost 1.73 from anaconda.

Obviously this isn't an RDKit bug, but I thought I'd mention it in case anyone else encounters the same issue.  

Dave

**Comment by bp-kelley** [comment]
I think I “fixed” this by changing the boost version, I’ll check my builds when I’m back at the keyboard.

**Reply by DavidACosgrove** [reply]
Hi Brian.  Thanks for looking at this.  I made a new conda/mamba build environment that pulled in boost 1.82 and it was all fine.
Cheers,
Dave

**Comment by odowd18** [comment]
Hello, I am running into issues incorporating rdkit into a C++ program on MacOS Sonoma14.3. I'm getting this error: 

I installed rdkit with conda and mamba.

n file included from /Users/xxxxxxxxxx/miniforge3/envs/rdkit_dev/include/rdkit/GraphMol/FileParsers/FileParsers.h:10:
In file included from /Users/xxxxxxxxxx/miniforge3/envs/rdkit_dev/include/rdkit/RDGeneral/export.h:3:
/Users/xxxxxxxxxx/miniforge3/envs/rdkit_dev/include/rdkit/RDGeneral/RDExportMacros.h:19:10: fatal error: 'boost/config.hpp' file not found
#include <boost/config.hpp>
         ^~~~~~~~~~~~~~~~~~

I am able to install rdkit and import it from python, but when I try to compile a C++ program with rdkit, I get this error on the Mac.

I tried the above with using boost 1.82 in my conda env, but it didn't help (Not really sure if I did it right there)

Would you have any suggestions?

Thank you

**Reply by greglandrum** [reply]
@giwds how are you compiling your program? You will need to make sure the compiler knows how to find both the boost header files (these are in $CONDA_PREFIX/include) and the RDKit header files

**Reply by odowd18** [reply]
The project uses make and cmake. 

I used find_package(RDKit REQUIRED) and linked    RDKit::SmilesParse  and RDKit::GraphMol
 in the CMakeLists.txt. Then I run make to build the program. This worked on my Linux machine.
I also installed rdkit with conda and mamba on my Linux machine.

I tried to add $CONDA_PREFIX/include with include_directories(...) in CMake, but no idea if that's the right way to do it. I also tried 
set(CMAKE_PREFIX_PATH ${CONDA_PREFIX}/include). Unfortunately, these did not solve the problem.

Apologies if I have some basic misunderstandings. I'm not familiar with make/cmake, etc.

**Reply by rvianello** [reply]
The boost package in conda-forge was split into multiple sub-packages at some point during the past several months. In a project of mine, where I use find_package(RDKit REQUIRED) for both linux and mac, I was required to update the configuration for the conda environment and make sure that libboost-devel was also installed (it's not automatically included as a runtime dependency for RDKit).

**Comment by DavidACosgrove** [comment]
@giwds this is the cmake command I use with success on Ventura 14.3 and which also worked on 14.1.
```
cmake -DPy_ENABLE_SHARED=1 \
      -DBOOST_ROOT="$CONDA_PREFIX" \
      -DBoost_NO_SYSTEM_PATHS=ON \
      -DBoost_NO_BOOST_CMAKE=TRUE \
      -DRDK_INSTALL_INTREE=ON \
      -DRDK_INSTALL_STATIC_LIBS=OFF \
      -DRDK_BUILD_CPP_TESTS=ON \
      -DPYTHON_NUMPY_INCLUDE_PATH="$(python -c 'import numpy ; print(numpy.get_include())')" \
      -DRDK_BUILD_CAIRO_SUPPORT=ON \
      -DRDK_BUILD_FREETYPE_SUPPORT=ON \
      ..
```
It has served me well over the years.  My mamba build environment was set up following the installation instructions, and its YAML file is below.  Probably some of my cmake options aren't necessary but I don't change what isn't broken.  At least one of the boost ones is essential.  Note that if you are using mamba rather than conda, it defaults to conda-forge and one of the python packages in the installation instructions isn't relevant.  I forget which, but it does no harm if you follow the instructions - it complains about the missing one but things still work, IIRC.


```
name: RDKit.Build.2
channels:
  - conda-forge
dependencies:
  - boost=1.82.0=h37858a2_6
  - boost-cpp=1.82.0=hca5e981_6
  - brotli=1.1.0=hb547adb_1
  - brotli-bin=1.1.0=hb547adb_1
  - bzip2=1.0.8=h3422bc3_4
  - c-ares=1.21.0=h93a5062_0
  - ca-certificates=2023.7.22=hf0a4a13_0
  - cairo=1.18.0=hd1e100b_0
  - certifi=2023.7.22=pyhd8ed1ab_0
  - cmake=3.27.6=h1c59155_0
  - contourpy=1.2.0=py312h76e736e_0
  - cycler=0.12.1=pyhd8ed1ab_0
  - eigen=3.4.0=h1995070_0
  - expat=2.5.0=hb7217d7_1
  - font-ttf-dejavu-sans-mono=2.37=hab24e00_0
  - font-ttf-inconsolata=3.000=h77eed37_0
  - font-ttf-source-code-pro=2.038=h77eed37_0
  - font-ttf-ubuntu=0.83=hab24e00_0
  - fontconfig=2.14.2=h82840c6_0
  - fonts-conda-ecosystem=1=0
  - fonts-conda-forge=1=0
  - fonttools=4.44.0=py312he37b823_0
  - freetype=2.12.1=hadb7bae_2
  - gettext=0.21.1=h0186832_0
  - icu=73.2=hc8870d7_0
  - kiwisolver=1.4.5=py312h389731b_1
  - krb5=1.21.2=h92f50d5_0
  - lcms2=2.15=hf2736f0_3
  - lerc=4.0.0=h9a09cb3_0
  - libblas=3.9.0=19_osxarm64_openblas
  - libboost=1.82.0=h72cdd8a_6
  - libboost-devel=1.82.0=hf450f58_6
  - libboost-headers=1.82.0=hce30654_6
  - libboost-python=1.82.0=py312h260253b_6
  - libboost-python-devel=1.82.0=py312h37858a2_6
  - libbrotlicommon=1.1.0=hb547adb_1
  - libbrotlidec=1.1.0=hb547adb_1
  - libbrotlienc=1.1.0=hb547adb_1
  - libcblas=3.9.0=19_osxarm64_openblas
  - libcurl=8.4.0=h2d989ff_0
  - libcxx=16.0.6=h4653b0c_0
  - libdeflate=1.19=hb547adb_0
  - libedit=3.1.20191231=hc8eb9b7_2
  - libev=4.33=h642e427_1
  - libexpat=2.5.0=hb7217d7_1
  - libffi=3.4.2=h3422bc3_5
  - libgfortran=5.0.0=13_2_0_hd922786_1
  - libgfortran5=13.2.0=hf226fd6_1
  - libglib=2.78.0=h24e9cb9_0
  - libiconv=1.17=he4db4b2_0
  - libjpeg-turbo=3.0.0=hb547adb_1
  - liblapack=3.9.0=19_osxarm64_openblas
  - libnghttp2=1.55.1=h2b02ca0_0
  - libopenblas=0.3.24=openmp_hd76b1f2_0
  - libpng=1.6.39=h76d750c_0
  - libsqlite=3.44.0=h091b4b1_0
  - libssh2=1.11.0=h7a5bd25_0
  - libtiff=4.6.0=ha8a6c65_2
  - libuv=1.46.0=hb547adb_0
  - libwebp-base=1.3.2=hb547adb_0
  - libxcb=1.15=hf346824_0
  - libzlib=1.2.13=h53f4e23_5
  - llvm-openmp=17.0.4=hcd81f8e_0
  - matplotlib=3.8.1=py312h1f38498_0
  - matplotlib-base=3.8.1=py312hba9b818_0
  - munkres=1.1.4=pyh9f0ad1d_0
  - ncurses=6.4=h463b476_2
  - numpy=1.26.0=py312h696b312_0
  - openjpeg=2.5.0=h4c1507b_3
  - openssl=3.1.4=h0d3ecfb_0
  - packaging=23.2=pyhd8ed1ab_0
  - pcre2=10.40=hb34f9b4_0
  - pillow=10.1.0=py312hac22aec_0
  - pip=23.3.1=pyhd8ed1ab_0
  - pixman=0.42.2=h13dd4ca_0
  - pkg-config=0.29.2=hab62308_1008
  - pthread-stubs=0.4=h27ca646_1001
  - pyparsing=3.1.1=pyhd8ed1ab_0
  - python=3.12.0=h47c9636_0_cpython
  - python-dateutil=2.8.2=pyhd8ed1ab_0
  - python_abi=3.12=4_cp312
  - readline=8.2=h92ec313_1
  - rhash=1.4.4=hb547adb_0
  - setuptools=68.2.2=pyhd8ed1ab_0
  - six=1.16.0=pyh6c4a22f_0
  - tk=8.6.13=hb31c410_0
  - tornado=6.3.3=py312h02f2b3b_1
  - tzdata=2023c=h71feb2d_0
  - wheel=0.41.3=pyhd8ed1ab_0
  - xorg-libxau=1.0.11=hb547adb_0
  - xorg-libxdmcp=1.1.3=h27ca646_0
  - xz=5.2.6=h57fd34a_0
  - zlib=1.2.13=h53f4e23_5
  - zstd=1.5.5=h4f39d0f_0
```

**Comment by odowd18** [comment]
Thank you @greglandrum @DavidACosgrove @rvianello 

It's now working.

I created a new conda env using the yml from Dave. It seems that Dave's yml has the libboost-devel which was mentioned by rvianello
Then I installed rdkit: `mamba install -c conda-forge cmake rdkit eigen`

Running make now compiles the program.

**Reply by thegodone** [reply]
Can you provide a full setup to make the compilation please?  

I am on 14.3.1 version of Sonoma

I was able to run the scripts but I have few segfault errors in the ctest maybe due to boost dynamic lib.

Thanks @bingodowd18

**Reply by odowd18** [reply]
Hi @thegodone, for my CMakelists.txt I used the following RDKit related settings:

```
cmake_minimum_required(VERSION 3.5)
set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED True)
find_package(RDKit REQUIRED)
...
set (RDKIT_LIBRARIES    RDKit::SmilesParse  RDKit::GraphMol)
```

and then the other stuff in my CMakeLists is for non-RDKit stuff.

I'm not sure if this is what you're looking for, but I hope it is helpful

**Comment by thegodone** [comment]
No this was not my question, sorry. In fact, I want to build rdkit (a
custom version) itself and I follow your instructions but I have those
errors at the ctest
```
cmake -DPy_ENABLE_SHARED=1 \
      -DRDK_USE_FLEXBISON=ON \
      -DBOOST_ROOT="$CONDA_PREFIX" \
      -DBoost_NO_SYSTEM_PATHS=ON \
      -DBoost_NO_BOOST_CMAKE=TRUE \
      -DRDK_INSTALL_INTREE=ON \
      -DRDK_INSTALL_STATIC_LIBS=OFF \
      -DRDK_BUILD_CPP_TESTS=ON \
      -DPYTHON_NUMPY_INCLUDE_PATH="$(python -c 'import numpy ;
print(numpy.get_include())')" \
      -DRDK_BUILD_CAIRO_SUPPORT=ON \
      -DRDK_BUILD_FREETYPE_SUPPORT=ON \
      -DCMAKE_BUILD_TYPE=Debug \
      ..


make -j12 & make install

export RDBASE=/Users/tgg/Github/rdkit-fir
export PYTHONPATH=$RDBASE
export DYLD_LIBRARY_PATH=${RDBASE}/lib

ctest —output_on_failure
```

The following tests FAILED:

  1 - testCoordGen (SEGFAULT)

34 - graphmoltestPickler (SEGFAULT)

51 - molbundleTestsCatch (SEGFAULT)

86 - testEnumeration (SEGFAULT)

89 - pyChemReactionEnumerations (Bus error)

93 - tautomerQueryTestCatch (SEGFAULT)

94 - pyTautomerQuery (SEGFAULT)

99 - pyFilterCatalog (SEGFAULT)

101 - pyFragCatalog (Bus error)

151 - pyMolDraw2D (Bus error)

170 - substructLibraryTest (SEGFAULT)

171 - substructLibraryCatchTest (SEGFAULT)

172 - pySubstructLibrary (SEGFAULT)

182 - pyGraphMolWrap (Bus error)

199 - testScaffoldNetwork (SEGFAULT)

201 - pyScaffoldNetworkPickling (Bus error)

208 - generalizedSubstructCatch (SEGFAULT)

209 - pyGeneralizedSubstruct (SEGFAULT)

219 - pythonProjectTests (Failed)
220 - pythonSourceTests (Failed)

Unfortunately it is the same on 2 different machines M1 (fresh
installation) and M3 and also on a clean computer.

in your yml, I did not see catch2 installation so I add it. I also need to
add flex and bison in my case but the error maybe related to boost.

I have also test on the last rdkit git commit and got the same error.

I don't see why I have those errors.

best


Le mar. 5 mars 2024 à 09:04, bingodowd18 ***@***.***> a
écrit :
